File: Java/JDI/jdi-uitest-core/src/main/java/com/epam/jdi/uitests/core/interfaces/complex/ISelector.java
package com.epam.jdi.uitests.core.interfaces.complex;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */

import com.epam.jdi.uitests.core.annotations.JDIAction;
import com.epam.jdi.uitests.core.interfaces.base.IBaseElement;
import com.epam.jdi.uitests.core.interfaces.base.ISetValue;

import java.util.List;

import static com.epam.commons.PrintUtils.print;

/**
 * Created by Roman_Iovlev on 6/10/2015.
 */
public interface ISelector<TEnum extends Enum> extends IBaseElement, ISetValue {
    /**
     * @param name Specify name using string
     *             Select Element with name (use text) from list
     */
    @JDIAction
    void select(String name);

    /**
     * @param name Specify name using enum
     *             Select Element with name (use enum) from list
     */
    @JDIAction
    void select(TEnum name);

    /**
     * @param index Specify digit to select
     *              Select Element with name (use index) from list
     */
    @JDIAction
    void select(int index);

    /**
     * @return Get name of the selected Element
     */
    @JDIAction
    String getSelected();

    /**
     * @return Get index of the selected Element
     */
    @JDIAction
    int getSelectedIndex();

    /**
     * @param name Specify name using string
     * @return Is option selected?
     */
    @JDIAction
    boolean isSelected(String name);

    /**
     * @param name Specify name using enum
     * @return Is option selected?
     */
    @JDIAction
    boolean isSelected(TEnum name);

    /**
     * @param name Specify name using string
     * Wait while option (from text) is selected. Return false if this not happens
     */
    @JDIAction
    void waitSelected(String name);

    /**
     * @param name Specify name using enum
     * Wait while option (from enum) is selected. Return false if this not happens
     */
    @JDIAction
    void waitSelected(TEnum name);

    /**
     * @return Get labels of all options
     */
    @JDIAction
    List<String> getOptions();

    default List<String> getNames() {
        return getOptions();
    }

    default List<String> getValues() {
        return getOptions();
    }

    /**
     * @return Get all options labels in one string separated with “; ”
     */
    @JDIAction
    default String getOptionsAsText() {
        return print(getOptions());
    }
}

File: C#.Net/JDI/Core/UnitTestProject1/Interfaces/Complex/ISelector.cs
using System;
using System.Collections.Generic;
using JDI_Core.Interfaces.Base;

namespace JDI_Core.Interfaces.Complex
{
    public interface ISelector : ISelector<IConvertible> { }

    public interface ISelector<in TEnum> : IBaseElement, ISetValue
        where TEnum : IConvertible
    {
        /**
     * @param name Specify name using string
     *             Select WebElement with name (use text) from list
     */
        //TODO[JDIAction]
        void Select(string name);

        /**
         * @param name Specify name using enum
         *             Select WebElement with name (use enum) from list
         */
        //TODO[JDIAction]
        void Select(TEnum name);

        /**
         * @param index Specify digit to select
         *              Select WebElement with name (use index) from list
         */
        //TODO[JDIAction]
        void Select(int index);

        /**
         * @return Get name of the selected WebElement
         */
        //TODO[JDIAction]
        string Selected();
        /**
         * @return Get index of the selected WebElement
         */
        //TODO[JDIAction]
        int SelectedIndex();

        /**
         * @param name Specify name using string
         * @return Is option selected?
         */
        //TODO[JDIAction]
        bool Selected(string name);

        /**
         * @param name Specify name using enum
         * @return Is option selected?
         */
        //TODO[JDIAction]
        bool Selected(TEnum name);

        /**
         * @param name Specify name using string
         * Wait while option (from text) is selected. Return false if this not happens
         */
        //TODO[JDIAction]
        void WaitSelected(string name);

        /**
         * @param name Specify name using enum
         * Wait while option (from enum) is selected. Return false if this not happens
         */
        //TODO[JDIAction]
        void WaitSelected(TEnum name);

        /**
         * @return Get labels of all options
         */
        //TODO[JDIAction]
        IList<string> Options { get; }

        //TODO GetName() was default
        IList<string> Names { get; }

        //TODO GetValues() was default 
        IList<string> Values { get; }

        /**
         * @return Get all options labels in one string separated with “; ”
         */
        //TODO[JDIAction]
        //TODO GetOptionsAsText() was default
        string OptionsAsText { get; }
    }
}
